README.md
# JupyFluo

JupyFluo is a Jupyter Notebook to analyze X-Ray Fluorescence (XRF) experiments on the beamline SIRIUS at the synchrotron SOLEIL.  
The notebook should be first set up by an Expert following the instructions in the "Expert" section. User can then follow the guidelines in the "User" section to start using the notebook. Please note that the notebook is currently in development. As such, be skeptical about any unexpected results.  Any feedback on the notebook or the code is welcome.

## Latest version
v0.5

## User

### Getting Started
1. Once the notebook is open, check that the following parameters in the second cell are correct:
```
# Name of the notebook: expt.notebook_name
# Directory where the data will be saved: expt.working_dir
# Directory where the nexus files are: expt.recording_dir
```  

2. Run the two first cells, check that no missing file is reported.

3. Use the dropdown list to choose the scan. Click on ```OK```.  

4. You can enter information about the scan by clicking on ```Insert comment```. 

### Extraction

1. Click on ```Extract the scan```.

2. Here you can define the channels of interest, and the spectrums you want to extract. Put first, for example, ```First Spectrum=0``` and ```Last Spectrum=1```. Click on ```Extract``` .

3. Use the figures to help you choose your subsets of channels and spectrums. 

4. Click again on ```Extract``` to update the parameters or click on ```Continue```.

5. A csv file ```Parameters_extraction.csv``` is created in the folder ```working_directory/filename/```.

### Set the fit parameters

1. Click on ```Set parameters```.

2. The parameters here should be already set by the expert user. You can still play with them and tick/untick the parameters you would like to fit.

3. Click on ```Validate```.

### Choose the peaks
1. Click on ```Set peaks```.  

2. A few peaks are already displayed (typically the Rayleigh and Compton peaks). Check their energies.

3. Use the tool ```Add peaks from database``` to import new peaks.

4. Modify the sheet to remove peaks or change their parameters. You can leave a peak in the list and do not include it in the analysis by writting ```no``` in the column ```#Fit Peak?```. 

5. **The strength of a line is its X-ray fluorescence production cross section with full cascade (in cm^2/g).** It is extracted from the database but can be obtained here as well http://lvserver.ugent.be/xraylib-web/.

6. Keep the peak/line names ```Elastic/El``` and ```Compton/Co```for the elastic (Rayleigh) and Compton peaks. You can add a Compton peak for an element by naming it ```Compton``` as well, and adding the element in the line column. For example, if you want to add a Compton peak for the Au La1 line, name the peak ```Compton``` and the line ```CoAuLa1```.

7. If you add an escape peak, do not name it with the name of its corresponding element! For example, if you add the escape peak of Au La1, name it 'EscAuLa1', name the line 'Esc', and put a strength of 1.

8. You can use the plot below the sheet to find where the peaks are. 

9. When you think you are done with the peaks, validate the sheet by clicking on ```Validate Peaks```.

10. Check the peaks on the spectrum. Modify them, or click on ```Start Fit``` to start the fit.

11. Note: You can also directly edit the Excel file ```Parameters_peaks.csv``` in your folder ```working_directory/filename/``` if you prefer.
 

### Fit the spectrums

When you click on ```Start Fit``` , you can follow the fit in real time. The results are continuously updated in your folder ```working_directory/filename/```. Check the files ```FitResult.csv``` and  the folder ```FitSpectrums/``` that you can open with you prefered software (Excel, Origin, Matlab, vim ...).


Once the fit is done, the fitted parameters will be displayed and saved as png in your folder ```working_directory/filename/```.
 
**The control panel will appear at the bottom of the plots when the fit is done.** 

### Add a plot to the PDF report

1. Until now, nothing will be rendered in the final PDF. Click on the button ```Add a plot to report```.

2. Choose a spectrum (if you were not working on the spectrums sum). Click on ```Preview the selected plot``` to preview it.

3. Choose the spectrum that you want to add to the PDF. Click on ```Add the selected plot```.

4. The output of the cell ```FF.Load_results(expt, spectrum_index=XXX)```, which is automatically generated and executed, will appear in the PDF.

5. Click on ```Export to pdf``` in the panel to generate the PDF, or ```Analyze a new scan``` to continue with the next scan. 

## Expert

### Getting Started

Start with a fresh download from the last main version on GitHub. If you rename the notebook, do not forget to change the corresponding parameter inside its first cell. Note that the notebook does not work with JupyterLab in its current version.

The aim of the Expert part is to determine the general and fitting parameters. It can be quite painful, but those parameters should not vary during an experiment. 

### General parameters for extraction

1. Click on  ```Extract```.

2. Fill in  ```First/Last channel```  and click on the right fluo elements.

3. Choose a range of channels covering from the lowest peak you want to fit to the end of the elastic peak (if included).

4. Extract about 10 spectrums from the file, and try to choose good ones (without any weird behaviour). 

### Fit parameters

**You should always start with a set of parameters already completed for the same energy as yours.** To do so, click on ```Load fit params``` after extracting the spectrums.   
Do not start from zero! Taking a previous set of parameters should directly give you reasonnable results.


### Determine the fit parameters
See inside ```lib/fit/```  and the section ```Quick description of the parameters``` of this file to have an explanation on the peak fitting. Here we detail a procedure which seems to be robust for determining the fit parameters, **when starting from a previous configuration.**

1. So far these parameters could be kept at zero and did not have to be fitted:

```
tfb0 = 0. (do not cancel twc0 at the same time)
fB = 0.
gammaB = 0.
```
 
2. Linear background:  
Put ```sl=0.```. Start by finding manually a reasonnable value for ```ct```, which should be close to the value of the background in a region without peaks.

3. Step function at low energy (<8 keV):  
Adjust ```sfa0``` manually (or you can try to fit it). It will change the step function at low energy.

4. Compton peaks:  
- Fit only ```gammaA```, which sets the slope of the foot of the Compton peaks at low energy. When the fit is done, get the average value using the button ```Extract averages```.
- Fit only ```fA```, which sets the transition between a linear slope to a Gaussian in the Compton foot, at low energy.

5. Gaussians width:  
Fit only ```noise``` and ```fG```.

6. Other parameters:  
If the fit is still not good enough, you can try to fit simulatenously ```sl``` and ```ct```.

### Peak definition

Determine the peaks to be used (see guidelines in the User section if needed).

### Finishing
1. Try to keep only ```ct``` as a fitting parameters for the User (or no fit parameters at all). If the Compton peak varies a lot with time, add ```gammaA```.

2. **Make the list of peaks the default one by clicking on ```Save peaks params default``` in the control panel.**

3. **Make the list of parameters the default one by clicking on ```Save fit params as default``` in the control panel.**

4. Delete the cells you have generated (keep only the first one), and the file is ready for the User.


## Description of the parameters

### References
- The spectrum function and peak definition are adapted from results in the following papers (please cite these references accordingly):
    -  M. Van Gysel, P. Lemberge & P. Van Espen,
    “Description of Compton peaks in energy-dispersive  x-ray fluorescence spectra”,
    X-Ray Spectrometry 32 (2003), 139–147
    -   M. Van Gysel, P. Lemberge & P. Van Espen, “Implementation of a spectrum fitting
    procedure using a robust peak model”, X-Ray Spectrometry 32 (2003), 434–441
    - J. Campbell & J.-X. Wang, “Improved model for the intensity of low-energy tailing in
    Si (Li) x-ray spectra”, X-Ray Spectrometry 20 (1991), 191–197  
    

- See this reference as well for results using these peak/spectrum definitions:
    - F. Malloggi, S. Ben Jabrallah, L. Girard, B. Siboulet, K. Wang, P. Fontaine, and J. Daillant, "X-ray Standing Waves and Molecular Dynamics Studies of Ion Surface Interactions in Water at a Charged Silica Interface", J. Phys. Chem. C, 123 (2019), 30294–30304  
    

- The  X-ray fluorescence production cross section are extracted from the database xraylib (https://github.com/tschoonj/xraylib/wiki):
    - T. Schoonjans et al. "The xraylib library for X-ray-matter interactions. Recent developments" Spectrochimica Acta Part B: Atomic Spectroscopy 66 (2011), pp. 776-784
(https://github.com/xraypy/XrayDB).

### Quick description of the parameters
Here we give a quick description of each parameters and typical values. For more details, see corresponding publications and the code itself.  

Note that for synchrotron-based experiments some parameters can be significantly different than parameters used with lab X-ray sources (most of them can actually be cancelled). 

- ```First/last channel```: the subset of channels to be extracted. Between 0 and 2047.
- ```First/last spectrum```: the subset of spectrums to be extracted. The first spectrum in the file has the index 0.


- ```Elements XXX```: check the box corresponding the detector elements. Typical value: ```Element 4``` for the single-element detector, ```Element 0,1,2,3``` for the four-elements detector.

- ```Energy```: the beamline energy in eV, for calculating the fluorescence cross sections. 

- ```sl, ct```: linear baseline ```ct+sl*eV```.
- ```noise```: electronic noise (FWHM in keV). Typical value: ```noise=0.1```.
- ```fG```: broadening factor of the gaussian width for the Compton peak. Typical value:  ```fG=1.-1.5```.


- ```sfa0```:  shelf fraction. Define the step function at low energy (~<8 keV). Typical value: ```sfa0=0.1-1.```.

- ```tfb0```: tail fraction. Can be cancelled. Typical value: ```tb0=0.```.  

- ```twc0```:  tail widths. Can be cancelled. Typical values: ```twc0=0.```.

- ```epsilon```: energy to create a charge carrier pair in the detector (in keV). Typical value for Si: ```epsilon=0.0036```.

- ```fano```: fano factor. Typical value for Si: ```fano=0.115```. 


- ```fA, fB```: Tail fractions fA (low energy) and fB (high energy) for the Compton peak.  Set the transition between a linear slope to a Gaussian in the Compton feet. Typical values: ```fA=0.05-0.3, fB=0.```.

- ```gammaA, gammaB```: Tail gammas gammaA (low energy) and gammaB (high energy). Set the slope of the feet of Compton peaks. Typical values: ```gammaA=1.-5., gammaB=0.```.  


- ```gain, ev0```: linear conversion channels/eVs through ```eVs = gain*channels + eV0```. Typical values: ```gain=9.9, ev0=0```.


- ```Limit iter.```: number of iterations at which a fit is considered as stuck and returns NaN values. Typical value: ```1000``` (but can be increased if needed).


- ```Use ipysheet```: use ipysheet or only Peak.csv to define the peaks. Typical value: ```True```.
- ```Fast extract```: use the fast extract option of PyNexus. Typical value: ```True```.
- ```Transmit fit params```: when ```True```, the results of a fit are the initial guess of the next fit. Can trigger a bad behaviour when there are sudden jumps in the spectrum.

- ```Set peaks on sum```: use the integration of all the sprectrums to define the peaks.

- ```Show zooms?```: show the zoom of each peak (takes time).  

- ```Show peaks?```: show the peaks or not in the plots.  

- ```Strength min```: minimal strength to appear in the peak database (in cm^2/g). 0.05 seems to be a reasonnable value. Put to zero to see all the peaks.

- The peak strengths are the X-ray fluorescence production cross section with full cascade (in cm^2/g). They are used to fix the relative intensity of each peak from the same atom.   

## Contributions
Contributions are more than welcome. Please report any bug or submit any new ideas to arnaud.hemmerle@synchrotron-soleil.fr or directly in the Issues section of the GitHub.


## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details


RELEASE_NOTES.md
# Release notes

### 03/06/2021  

Update to v0.5:  
- Rewrite the software architecture to avoid extraction of the scan every time a parameter is changed

### 02/06/2021  

Update to v0.4:  
- Massive simplification of the fit procedure (less parameters)
- Link to the library of JupyLabBook (XRR.py)
- Definition of a Roadmap for v1.0 


lib/frontend/Action.py
from . import Utils
from lib.extraction.common import PyNexus as PN

import os
import numpy as np
import matplotlib.pyplot as plt
import ipywidgets as widgets
import time
import subprocess
import math

import shutil

"""Frontend library for all the widgets concerning the actions in the notebook."""

    
def Check_and_init(expt):
    '''
    Check if files and folders exist, then create the first cell.

    Parameters
    ----------
    expt : object
        object from the class Experiment
    '''

    print("Results will be saved in the folder:")
    if not os.path.exists(expt.working_dir):
        print(PN._RED+"Careful, the following folder does not exist and should be created:"+PN._RESET)
    print(expt.working_dir)
    print("")
        
    print("The original nexus files should be in the folder:")    
    if not os.path.exists(expt.recording_dir):
        print(PN._RED+"Careful, the following folder does not exist:"+PN._RESET)
    print(expt.recording_dir)
    print("")
    
    if not os.path.exists(expt.notebook_name):
        print(PN._RED+"Careful, assign the correct notebook name to expt.notebook_name."+PN._RESET)
        print("")

    file = 'parameters/peaks/Default_Parameters_peaks.csv'
    if not os.path.exists(file):
        print(PN._RED+"The following file does not exist:"+PN._RESET)
        print(file)
        print("This file contains the peaks to be displayed by default.")
        print("")
     
    file = 'parameters/extraction/Default_Parameters_extraction.csv'
    if not os.path.exists(file):
        print(PN._RED+"The following file does not exist:"+PN._RESET)
        print(file)
        print("This file contains the parameters for extraction by default.")
        print("")

    file = 'parameters/fit/Default_Parameters_fit.csv'
    if not os.path.exists(file):
        print(PN._RED+"The following file does not exist:"+PN._RESET)
        print(file)
        print("This file contains the parameters for extraction by default.")
        print("")        
        
    Utils.Create_cell(code='FE.Action.Choose(expt)', position ='at_bottom', celltype='code', is_print=False)

def Choose(expt):
    '''
    Choose the next action to do.

    Parameters
    ----------
    expt : object
        object from the class Experiment
    '''

    # Define the list of nxs files in the recording directory
    expt.list_nxs_files = [file for file in sorted(os.listdir(expt.recording_dir)) if 'nxs' in file][::-1]
    if expt.list_nxs_files == []:
        print(PN._RED+'There is no nexus file in the recording folder.'+PN._RESET)
        print(PN._RED+'Recording folder: %s'%expt.recording_dir+PN._RESET)
        expt.list_nxs_files = ['SIRIUS_NoFileFound_00_00_00.nxs']
        
    def selection_scan(nxs_file):
        """
        Called by the widget to select the scan to be treated.
        """

        # Generate several identifiers for the scan
        expt.nxs = nxs_file
        Define_scan_identifiers(expt)
        
        # Create a folder for saving params and results, if it does not already exist.
        if not os.path.exists(expt.working_dir+expt.id):
            os.mkdir(expt.working_dir+expt.id)

        # Check if the csv files for parameters already exists, if not copy the default ones
        if not os.path.isfile(expt.working_dir+expt.id+'/Parameters_extraction.csv'):
            shutil.copy('parameters/extraction/Default_Parameters_extraction.csv',
                        expt.working_dir+expt.id+'/Parameters_extraction.csv')

        if not os.path.isfile(expt.working_dir+expt.id+'/Parameters_fit.csv'):
            shutil.copy('parameters/fit/Default_Parameters_fit.csv',
                        expt.working_dir+expt.id+'/Parameters_fit.csv')            
            
    def on_button_treat_clicked(b):
        """
        Generate and execute cells corresponding to the chosen scan.
        """
        #clear_output(wait=False)
        
        selection_scan(w_print_scan.value)
        
        Utils.Create_cell(code='FE.Treatment.Choose(expt)',
                    position ='below', celltype='code', is_print=False)

    
        Utils.Create_cell(code='### '+expt.id, position ='below', celltype='markdown', is_print=True)
        
        Utils.Delete_current_cell()
        
    def on_button_refresh_clicked(b):
        """
        Re-execute the cell to update it.
        """
        
        Utils.Refresh_current_cell()
     
        
    def on_button_export_clicked(b):
        """
        Export the notebook to PDF.
        """
        
        print('Export in progress...')
        
        export_done = Export_nb_to_pdf(expt.notebook_name)
        
        if export_done:
            print('Notebook exported to %s.pdf'%expt.notebook_name.split('.')[0])
        else:
            print("There was something wrong with the export to pdf.")
            print("Did you rename the Notebook? If yes:")
            print("1) Change the value of expt.notebook_name in the first cell (top of the Notebook).")
            print("2) Re-execute the first cell.")
            print("3) Try to export the pdf again in the last cell (bottom of the Notebook).")

                
    def on_button_markdown_clicked(b):
        """
        Insert a markdown cell below the current cell.
        """ 
        
        Utils.Delete_current_cell()
        
        Utils.Create_cell(code='', position ='below', celltype='markdown', is_print=True, is_execute=False)
    
        Utils.Create_cell(code='FE.Action.Choose(expt)', position ='at_bottom', celltype='code', is_print=False)
        
    
    # Display the widgets
   
    # Click to treat a single scan
    button_treat = widgets.Button(description="Treat scan")
    button_treat.on_click(on_button_treat_clicked)
    
    # Click to refresh the list of files
    button_refresh = widgets.Button(description="Refresh")
    button_refresh.on_click(on_button_refresh_clicked)
    
    # Click to export to pdf
    button_export = widgets.Button(description="Export to PDF")
    button_export.on_click(on_button_export_clicked)
    
    # Click to insert a markdown cell
    button_markdown = widgets.Button(description="Insert comment")
    button_markdown.on_click(on_button_markdown_clicked)

    # Widget for selection of scan
    w_print_scan = widgets.Dropdown(
                    options=expt.list_nxs_files,
                    layout=widgets.Layout(width='400px'),
                    style={'description_width': 'initial'})
                    
    buttons0 = widgets.HBox([w_print_scan, button_treat])
    display(buttons0)
                    
    buttons1 = widgets.HBox([button_refresh, button_export, button_markdown])
    display(buttons1)
    

def Export_nb_to_pdf(nb_name):
    '''
    Export the notebook to pdf using a command line through the OS.

    Parameters
    ----------
    nb_name : str
        full name of the notebook. Ex: 'JupyLabBook.ipynb'

    Returns
    -------
    bool
        export_done, True if the export suceeded without error/warning
    '''
    
    # Save the current state of the notebook (including the widgets)
    Utils.Save_nb()
    
    t0 = time.time()
    rc = 1
    while rc>0:
        if (time.time()-t0) > 100:
            # Timeout before PDF export is considered as failed
            export_done = False
            break
        else:
            time.sleep(3)
            command = 'jupyter nbconvert '
            command+= nb_name
            command+= ' --to pdf '
            command+= ' --TagRemovePreprocessor.remove_cell_tags=\"[\'notPrint\']\" ' # Remove the widgets from the PDF
            command+= ' --no-input ' # Remove the code cells
            #command+= '--template latex_template.tplx' # Custom template (not working with nbconvert >= 6.0)
            rc = subprocess.call(command,shell=True)
            if rc==0: export_done = True
                
    return export_done




def Define_scan_identifiers(expt):
    '''
    Create a series of identifiers for the current scan.

    Parameters
    ----------        
    expt : object
        object from the class Experiment
    '''

    # For example:
    # expt.nxs = 'SIRIUS_2017_12_11_08042.nxs'
    # expt.path = '/Users/arnaudhemmerle/recording/SIRIUS_2017_12_11_08042.nxs'
    # expt.id = 'SIRIUS_2017_12_11_08042'
    # expt.number = 8042
    
    expt.path = expt.recording_dir+expt.nxs
    expt.id = expt.nxs[:-4]
    split_name = expt.nxs.split('.')[0].split('_')
    expt.number = int(expt.nxs.split('.')[0].split('_')[-1])
    
    # Sets the tokens
    expt.is_extract_done = False
    expt.is_set_peaks_done = False
    expt.is_fit_done = False
    



    

lib/fit/Functions.py
import numpy as np
from scipy.special import erfc


def Fcn_spectrum(dparams, groups, channels):
    '''
    Definition of the spectrum as the sum of the peaks + the background.

    Parameters
    ----------
    dparams : dict
        contains the parameters
    groups : array_like
        array of objects Group containing info on each peak
    channels : array_like
        list of channels

    Returns
    -------
    array_like
        spectrum_tot, the calculated spectrum
    array_like
        gau_tot, the contribution of the gaussian functions to the spectrum
    array_like
        she_tot, the contribution of the shelf functions to the spectrum
    array_like
        tail_tot, the contribution of the tail functions to the spectrum
    array_like
        baseline, the contribution of the baseline to the spectrum
    array_like
        compton, the contribution of the compton functions to the spectrum
    array_like
        eVs, the channels converted to eVs
    '''

    # Unpack
    eVs = dparams['gain']*channels + dparams['eV0']
    ct = dparams['ct']
    sl = dparams['sl']
    noise = dparams['noise']

    # The total fit for the whole spectrum
    spectrum_tot = 0.
    # The different parts composing the fit (for control)
    gau_tot = 0.
    she_tot = 0.
    tail_tot = 0.
    compton = 0.
    for group in groups:
        # The total fit for the group of peaks
        spectrum_group = 0.
        # The different parts composing the fit (for control)
        gau_group = 0.
        she_group = 0.
        tail_group = 0.   
        area = group.area
        for peak in group.peaks:
            position = peak.position
            intensity_rel = peak.intensity_rel
            if group.elem_name == 'Compton':
                compton += area*Fcn_compton_peak(position,intensity_rel,channels,dparams)
                spectrum_group += compton
            else:
                ppic, gau, she, tail = Fcn_peak(position,intensity_rel,channels,dparams)
                spectrum_group += area*ppic
                gau_group += area*gau
                she_group += area*she
                tail_group += area*tail
                compton += ppic*0.
                
        spectrum_tot += spectrum_group
        gau_tot += gau_group
        she_tot += she_group
        tail_tot += tail_group

    
    # Comment/uncomment to cut the baseline at the end of the elastic peak
    # Do not forget to comment the line baseline = ct+sl*eV
    
    # We add a linear baseline, which cannot be < 0, and stops after the elastic peak (if there is one)
    limit_baseline = eVs[-1]
    for group in groups:
        if group.elem_name == 'Elastic':
            for peak in group.peaks:
                if peak.name == 'El':
                    limit_baseline = peak.position                
    eVs_tmp = np.where(eVs<limit_baseline+1000*noise, eVs, 0.)
    baseline = ct+sl*eVs_tmp
    
    
    #baseline = ct+sl*eVs
    baseline = np.where(baseline>0.,baseline,0.)
    spectrum_tot+= baseline

    return spectrum_tot, gau_tot, she_tot, tail_tot, baseline, compton, eVs


def Interpolate_scf(atom, energy):
    '''
    Interpolation for the scattering factors.
    Used here to take into account absorption from Si within the detector.
    Requires the file f-si.    
    
    Parameters
    ----------
    atom : str
        the atom of interest (here 'Si')
    energy : float
        energy in eV
        
    Returns
    -------
    float
        scf, f1 (from CXRO)
    float
        scfp, f2 (from CXRO)
    '''

    en2=0.
    f2=0.
    f2p=0.
    for line in open('lib/fit/f-'+str(atom),'r'):
        en1=en2
        f1=f2
        f1p=f2p
        try:
            en2=float(line.split()[0])
            f2=float(line.split()[1])
            f2p=float(line.split()[2])
            if en1<=energy and en2>energy:
                scf=f1+(energy-en1)/(en2-en1)*(f2-f1)
                scfp=f1p+(energy-en1)/(en2-en1)*(f2p-f1p)
            else:
                pass
        except:
            pass
    return scf,scfp

def Fcn_peak(pos, amp, channels, dparams):
    '''
    Definition of a peak (area normalised to 1).
    Following:
    - M. Van Gysel, P. Lemberge & P. Van Espen, “Implementation of a spectrum fitting
    procedure using a robust peak model”, X-Ray Spectrometry 32 (2003), 434–441
    - J. Campbell & J.-X. Wang, “Improved model for the intensity of low-energy tailing in
    Si (Li) x-ray spectra”, X-Ray Spectrometry 20 (1991), 191–197
    The params for peak definition should be passed as a dictionnary :
    dparams = {'sl': 0.01, 'ct':-23., 'sfa0':1.3 ... }    
    
    Parameters
    ----------
    pos : float
        position of the peak in eVs 
    amp : float
        amplitude of the peak
    channels : array_like
        list of channels        
    dparams : dict
        contains the parameters
        
    Returns
    -------
    array_like
        ppic, the calculated intensity for each value of channels (model peak)
    array_like
        np.array(gau), the gaussian contribution to ppic
    array_like
        np.array(SF*she), the shelf contribution to ppic
    array_like
        np.array(TF*tail), the tail contribution to ppic
    '''

    # Unpack
    eVs = dparams['gain']*channels + dparams['eV0']
    
    #We rescale sfa0 to avoid very little value of the input
    sfa0 = dparams['sfa0']/1e4 
    tfb0 = dparams['tfb0']
    twc0 = dparams['twc0']
    noise = dparams['noise']
    fano = dparams['fano']
    epsilon = dparams['epsilon']

    # We neglect these contributions to the model, which do not 
    # seem necessary for synchrotron radiations
    sfa1 = 1e-15
    tfb1 = 1e-15
    twc1 = 1e-15
    
    # We work in keV for the peak definition
    pos_keV = pos/1000.
    keVs = eVs/1000.

    # Peak width after correction from detector resolution (sigmajk)
    wid = np.sqrt((noise/2.3548)**2.+epsilon*fano*pos_keV)

    # Tail width (cannot be <0)
    TW = twc0 + twc1*pos_keV
    TW = np.where(TW>0.,TW,0.)

    # Energy dependent attenuation by Si in the detector
    atwe_Si = 28.086 #atomic weight in g/mol
    rad_el = 2.815e-15 #radius electron in m
    Na = 6.02e23 # in mol-1
    llambda = 12398./pos*1e-10 # in m
    # mass attenuation coefficient of Si in cm^2/g
    musi = 2.*llambda*rad_el*Na*1e4*float(Interpolate_scf('si',pos)[1])/atwe_Si

    # Shelf fraction SF (cannot be <0)
    SF = (sfa0 + sfa1*pos_keV)*musi
    SF = np.where(SF>0.,SF,0.)
    
    # Tail fraction TF (cannot be <0)
    TF = tfb0 + tfb1*musi
    TF = np.where(TF>0.,TF,0.)

    # Definition of gaussian
    arg = (keVs-pos_keV)**2./(2.*wid**2.)
    farg = (keVs-pos_keV)/wid
    gau = amp/(np.sqrt(2.*np.pi)*wid)*np.exp(-arg)
    # Avoid numerical instabilities
    gau = np.where(gau>1e-10,gau, 0.)

    # Function shelf S(i, Ejk)
    she = amp/(2.*pos_keV)*erfc(farg/np.sqrt(2.))
    # Avoid numerical instabilities
    she = np.where(she>1e-10,she, 0.)

    # Function tail T(i, Ejk)
    tail = amp/(2.*wid*TW)*np.exp(farg/TW+1/(2*TW**2))*erfc(farg/np.sqrt(2.)+1./(np.sqrt(2.)*TW))
   
    # Avoid numerical instabilities
    tail = np.where(tail>1e-10,tail, 0.)

    # Function Peak
    ppic = np.array(gau+SF*she+TF*tail)

    return ppic, np.array(gau), np.array(SF*she), np.array(TF*tail)


def Fcn_compton_peak(pos, amp, channels, dparams):
    '''
    The function used to fit the compton peak, inspired by  M. Van Gysel, P. Lemberge & P. Van Espen,
    “Description of Compton peaks in energy-dispersive  x-ray fluorescence spectra”,
    X-Ray Spectrometry 32 (2003), 139–147
    The params for peak definition should be passed as a dictionnary :
    dparams = {'fG': 0.01, 'fA':2., ...}

    Parameters
    ----------
    pos : float
        position of the peak in eVs 
    amp : float
        amplitude of the peak
    channels : array_like
        list of channels        
    dparams : dict
        contains the parameters

    Returns
    -------
    array_like
        ppic, the calculated intensity for each value of channels (model peak)
    '''

    eVs = dparams['gain']*channels + dparams['eV0']
    fG = dparams['fG']
    noise = dparams['noise']
    fano = dparams['fano']
    epsilon = dparams['epsilon']

    # We work in keV for the peak definition
    pos_keV = pos/1000.
    keVs = eVs/1000.

    # Peak width after correction from detector resolution (sigmajk)
    wid = np.sqrt((noise/2.3548)**2.+epsilon*fano*pos_keV)

    # Definition of gaussian
    arg = (keVs-pos_keV)**2./(2.*(fG*wid)**2.)
    gau = amp/(np.sqrt(2.*np.pi)*fG*wid)*np.exp(-arg)
  
    fA = dparams['fA']
    fB = dparams['fB']
    gammaA = dparams['gammaA']
    gammaB = dparams['gammaB']
    

    #Low energy tail TA
    farg = (keVs-pos_keV)/wid
    if (fA<0.001 or gammaA<0.001):
        TA =0.
    else:
        TA = amp/(2.*wid*gammaA)*np.exp(farg/gammaA+1/(2*gammaA**2))*erfc(farg/np.sqrt(2.)+1./(np.sqrt(2.)*gammaA))

    #High energy tail TB
    if (fB<0.001 or gammaB<0.001):
        TB = 0.
    else:
        TB = amp/(2.*wid*gammaB)*np.exp(-farg/gammaB+1/(2*gammaB**2))*erfc(-farg/np.sqrt(2.)+1./(np.sqrt(2.)*gammaB))

    ppic = np.array(gau+fA*TA+fB*TB)
    
    # Avoid numerical instabilities
    ppic = np.where(ppic>1e-10,ppic, 0.)
    
    return ppic